policy-controller/src/admission.rs

```diff
@@ -72,31 +73,45 @@ fn ok(reply: impl warp::Reply + 'static) -> Result<Box<dyn warp::Reply>, Infalli
 }
 
 /// Parses a `Server` instance and it's namespace from the admission request.
-fn parse_server(req: Request) -> Result<(String, api::policy::Server)> {
+fn parse_server(req: Request) -> Result<(String, String, api::policy::ServerSpec)> {
     let obj = req.object.ok_or_else(|| anyhow!("missing server"))?;
-    let server = serde_json::from_value::<api::policy::Server>(obj.data)?;
-    let ns = server
+    let ns = obj
         .namespace()
         .ok_or_else(|| anyhow!("no 'namespace' field set on server"))?;
-    Ok((ns, server))
+    let name = obj.name();
+    let data = serde_json::from_value::<RequestData>(obj.data)?;
+    Ok((ns, name, data.spec))
 }
 
 /// Validates a new server (`review`) against existing `servers`.
-fn validate(review: &api::policy::Server, servers: &[api::policy::Server]) -> Result<()> {
+fn validate(
+    review_name: &str,
+    review_spec: &api::policy::ServerSpec,
+    servers: &[api::policy::Server],
+) -> Result<()> {
     for s in servers {
         // If the port and pod selectors select the same resources, fail the admission of the
         // server. Ignore existing instances of this Server (e.g., if the server's metadata is
         // changing).
-        if s.name() != review.name()
+        if s.name() != review_name
             // TODO(ver) this isn't rigorous about detecting servers that select the same port if one port
             // specifies a numeric port and the other specifies the port's name.
-            && s.spec.port == review.spec.port
+            && s.spec.port == review_spec.port
             // TODO(ver) We can probably detect overlapping selectors more effectively.
-            && s.spec.pod_selector == review.spec.pod_selector
+            && s.spec.pod_selector == review_spec.pod_selector
         {
             bail!("identical server spec already exists");
         }
     }
 
     Ok(())
 }
+
+/// RequestData is a wrapper around the `data` field of the AdmissionRequest.
+/// The data field contains a `serde_json::value::Value::Object` type that wraps
+/// the spec of the resource, deserializing directly to a ServiceSpec won't work
+/// in this case.
+#[derive(serde::Serialize, serde::Deserialize, Debug)]
```

Comment:
```suggestion
#[derive(serde::Deserialize, Debug)]
```

we don't actually need to serialize this do we?